apache/kafka
# KAFKA-16855: remote log disable policy in KRaft

Implementation for KIP-950. 
This PR did:
1. added `remote.log.copy.disabled` config, and remove leaderCopyTask and followerTask when enabled
2. added `remote.log.delete.on.disable` config, and fail the config change when `remote.storage.enable=false`, but `remote.log.delete.on.disable=false`.
3. remove all remote logs when `remote.storage.enable=false` and `remote.log.delete.on.disable=true`.
4. disallow enabling `remote.log.delete.on.disable` or `remote.log.copy.disabled` configs in ZK mode.
5. added unit tests
6. added integration test scenario:
```
1. create a topic with tiered enabled
2. produce some records and wait for offloading to remote storage
3. set `remote.log.copy.disabled` config
4. make sure the data is consumable from remote storage
5. re-enable the remote storage (set remote.log.copy.disabled=false)
6. produce some data and make sure data is offloaded
7. explicitly set `remote.log.copy.disabled` config
8. delete 1 record
9. make sure remote data is deleted (retention policy works well)
10. disable remote storage with `remote.log.delete.on.disable` set
11. make sure all remote storage data is deleted.

```

### Committer Checklist (excluded from commit message)
